ieeeuoft/hackathon-template
# Convert Cart Page and Components to Typescript, Connect CartCard to Redux

## Overview

Resolves IEEE-112

Converts the `CartCard` component and `Cart` page to Typescript, and connects the former to redux to grab hardware. This was more of a rewrite than a simple refactor, since the (implicit) interfaces we assumed for the cart originally don't match up with what we have decided (and can now enforce with Typescript).

The `CartCard` component now just takes a hardware ID and quantity, and pulls the hardware from the redux store. I wanted it to just take a regular `Hardware` object, but pulling the relevant objects from the store inside the `Cart` component was rather difficult and would have led to inefficient re-renders of the entire `Cart` component in some cases. The downside is that tests are a little harder to write this way.

A new `makeStoreWithEntities` test helper was added, which pre-fills the store with entities (currently only supports hardware and categories, more have to be added manually). This does rely a bit on knowing the internal workings of redux, but typescript will yell at us if that ever changes.

Finally, adds an argument to the `getHardwareWithFilters` thunk to allow preserving existing hardware in the store. This is used by the cart page if anything new needs to be fetched.

## Unit Tests Created

- Added a new `makeStoreWithEntities` helper function
- Updated the `Cart` and `CartCard` tests. Some tests were removed since they'll need to change with redux, comments left in their place.
- New test that the Cart page fetches missing hardware on load
- Some extra basic tests for the hardware and category thunks, to make using the new `makeStoreWithEntities` safer


## Steps to QA

- The current mocked cart items use hardware with IDs 1, 2, and 3, so make sure you have at least those added (via the admin site) otherwise weird stuff will happen. Then just visit the cart page.
